### ISSUE TITLE
Replace hard-coded user roles with fully configurable user roles and remove REST_USER user role [5]

### ISSUE THREAD

**Issue by MartinWahnschaffe** [hzi-braunschweig/SORMAS-Project#4461]
### Problem Description
The current user roles are hard-coded, but need to become manageable by the admin. There is already a first step made towards this in #830, which allows to override the user rights configuration in the database.

Source: https://github.com/hzi-braunschweig/SORMAS-Project/issues/898#issuecomment-671336289

### Proposed Change

Depends on #2804.

#### Preparation

- [x] Existing UserRole enum should become private DefaultUserRole enum and should be moved to sormas-backend
- [x] Move mapping between DefaultUserRole and UserRight from UserRight enum to DefaultUserRole enum

#### UserRole entity

- [x] The existing UserRoleConfig class needs to become the new UserRole class. When loading user roles only use the UserRole entries from the database (in other words: Don't mix-in user roles created on the fly based on the default user roles or something like that).
- [x] UserRole: Add caption and description fields
- [x] UserRole: Add hasOptionalHealthFacility, hasAssociatedOfficer and portHealthUser boolean fields. Those might be replaced with other mechanics in the future
- [x] UserRole: Add a list of notifications to the user role
- [x] User: Replace Set of old UserRole enum with Set of new UserRole entity. Will require a user:userRoles mapping table.
- [x] Provide a method to create user roles based on all default roles. Should be called similar to the default users when the system is set up
- [x] ~~Should we keep a defaultUserRole field to know that a user roles was created based on a default user role? Could be helpful for migrations later. Otherwise~~ remove the userRole field.
- [x] the getAssignableUserRoles logic can be dropped - it should actually already have been dropped with #659

#### Migration

- [x] When updating from a previous version call the method to create user roles based on the default user roles
- [x] Migration of user roles for existing users

#### Android App

- [x] Replace userRolesJson in User with join table
- [x] UserRoleConfig becomes UserRole
  - Create new table and drop the old one
  - Include all fields that have been added in #4461 to the UserRole. Notification types can be ignored
  - userRights can be stored as json
- [x] Change the UserDao.getUsersByAssociatedOfficer method (introduced by #8318) to retrieve useres based on the actual user rights (join user roles and do a "like" check on the user rights json)
- [x] Same for the logic needed for getRandomRegion/DistrictCaseResponsible introduced by #8318

Out of scope: Add translations for user role names

## Release Notes
The REST_USER default user role has been removed, because it has become obsolete after we had to remove the logic that has given users of this role access to data across all jurisdictions.
All default user roles now come with the SORMAS_REST user right, so for most use cases (sormas angular, external visits) this will not have any effect. For the survnet converter the user roles of the related user have to be changed to the NATIONAL_USER.

**Comment by Marko-ilmkreis** [comment]
Very good. It is highly needed also for data protection purposes. But I know a big thing to change it!

**Comment by JonasCir** [comment]
@MartinWahnschaffe Do you have any estimate when this will be available? :)

**Comment by MartinWahnschaffe** [comment]
@JonasCir There are a few dependencies that need to be done first: #2802, #2804, #2805

**Comment by MartinWahnschaffe** [comment]
@ChristopherRiedel I have taken the freedom to put the Android parts of this in here, because the other parts of this issue can't be merged before this is done. The last two points of this depends on #8318 - @MateStrysewske is now working on this.

**Comment by ChristopherRiedel** [comment]
UserRoleConfigs get truncated

**Comment by kwa20** [comment]
@MartinWahnschaffe @ChristopherRiedel How are we handling users that currently serve rather technical purposes and are not actually used by humans like the survnet user (not actually an independent role) or the climedo user (external visits user). Those users/roles should probably not be configurable to guarantee the respective functionality.

**Comment by AndyBakcsy-she** [comment]
From QA team in order to fully understand the implications of this ticket, we will need more information about what is needed to change and what was affected.

1. Is this ticket comment relevant for the editing of user roles? 
https://github.com/hzi-braunschweig/SORMAS-Project/issues/830#issuecomment-434306233

2. How will we be able to change configurations for user roles? (feature configuration or via DB changes?

**Comment by ChristopherRiedel** [comment]
> From QA team in order to fully understand the implications of this ticket, we will need more information about what is needed to change and what was affected.
> 
> 1. Is this ticket comment relevant for the editing of user roles?
>    [Make user rights of user roles configurable [1+3] #830 (comment)](https://github.com/hzi-braunschweig/SORMAS-Project/issues/830#issuecomment-434306233)
> 2. How will we be able to change configurations for user roles? (feature configuration or via DB changes?

1. It will not yet be possible to edit UserRoles via the UI after this issue. -> #4462 #4463
2. It is not necessarily intended that UserRoles are edited after this issue, since UserRoleConfigs are also dropped with this issue. However, it is of course possible to edit the UserRoles directly in the database, because after this issue all UserRoles are completely in the database.

**Comment by ChristopherRiedel** [comment]
> @MartinWahnschaffe @ChristopherRiedel How are we handling users that currently serve rather technical purposes and are not actually used by humans like the survnet user (not actually an independent role) or the climedo user (external visits user). Those users/roles should probably not be configurable to guarantee the respective functionality.

As discussed with @kwa20, we will not implement something like this for now. It is not possible to customize the UserRoles via the UI after this issue anyway, this will come with the issues #4462 and #4463.

**Comment by ChristopherRiedel** [comment]
@bernardsilenou @Chinedar 
Would it be a big problem if we would remove the userroles from user captions?
If yes, would it be okay to remove them now and to bring them back in the future?
I'm asking this because displaying them would lead to performance problems on showing lists of users.

**Comment by Chinedar** [comment]
It should not be a problem, if there is an alternative to identify what
rights a user still has. Or I'm I missing the point of the intervention?

**Comment by ChristopherRiedel** [comment]
@Chinedar 
> It should not be a problem, if there is an alternative to identify what rights a user still has. Or I'm I missing the point of the intervention?
In the user directory it would still be possible to read everything as before. This is about the captions of the users, which are displayed e.g. in user selection fields or user filters in the ui.

**Comment by Marko-ilmkreis** [comment]
@ChristopherRiedel @Chinedar from the view of the end.user it would be also very good to remove them. I did not see any benefit to see it there.

**Comment by Chinedar** [comment]
Then this is good to go from my side. Thanks.

**Comment by bernardsilenou** [comment]
@ChristopherRiedel 
Sorry to reply late, I have been on and off on the road.
- To me, I think removing the default user roles would make it very difficult to manage the users. The user accounts are often named by their roles and this is what we used during training sessions. These roles are not what each admin may like to create but were designed based on the IDSR system used in most countries <https://www.afro.who.int/publications/technical-guidelines-integrated-disease-surveillance-and-response-african-region-third>. 
- If removing the roles in this version and adding them back is the only way, then we can go ahead, but mention it in the release note so that countries that depend on the default user roles would not use the version.

**Comment by ChristopherRiedel** [comment]
@bernardsilenou 
> @ChristopherRiedel Sorry to reply late, I have been on and off on the road.
> 
> * To me, I think removing the default user roles would make it very difficult to manage the users. The user accounts are often named by their roles and this is what we used during training sessions. These roles are not what each admin may like to create but were designed based on the IDSR system used in most countries https://www.afro.who.int/publications/technical-guidelines-integrated-disease-surveillance-and-response-african-region-third.
> * If removing the roles in this version and adding them back is the only way, then we can go ahead, but mention it in the release note so that countries that depend on the default user roles would not use the version.

This is not about removing the default user roles.
It's about changing the captions of users e.g. in the responsible surveillance officer field.
The captions would change from the old pattern (Adam Smith - Surveillance Officer) to the new pattern (Adam Smith).

**Comment by bernardsilenou** [comment]
@ChristopherRiedel 
I think i now get what you mean, I was thinking something different. One more question please
I this the place that you mean ?
![image](https://user-images.githubusercontent.com/12996724/164224600-15334930-707b-4325-bfb7-96ab27ec832e.png)

if yes, and this dropdown would soon represents responsible users containing all user roles (just as in events), if I want to assign another surveillance officer as a responsible user, how would I easily know that Adam Smith is a surveillance officer?

**Comment by bernardsilenou** [comment]
@ChristopherRiedel 
I think my last comment is optional, so we can remove and add later

**Comment by ChristopherRiedel** [comment]
@bernardsilenou 
Yes that would be one of the places where the listing of user roles would be removed. The same would happen for user filters for example. The users would still be displayed, of course, but the user roles behind them would be temporarily dropped.

Regarding your optional comment: Exactly such cases are the reason why the listing should come back in the medium term.

**Comment by bernardsilenou** [comment]
Having the user role attached to the name is also very important during training. A surveillance officer for example is called DSNO (in Nigeria) and DCO (in Ghana). Having the role shown together with the name of the user helps the user to understand the sormas personas and who they are in terms of these personas.

**Comment by bernardsilenou** [comment]
@ChristopherRiedel  
I also agree to temporally remove the user roles depending on what we mean by temporal.
If it's something that we are not sure to add the user role to the version that follows, then this is the point that I do not agree. 
My suggestion is to remove it if it resolves the bottleneck and add it back in the subsequent release.
-The current solution shows all the user roles of the user, this is not practical if the user has many roles,  maybe when adding the user roles back, we can show only the dominant role, e.g if a user has the surveillance and contact officer roles, then show only surveillance officer.

**Comment by Marko-ilmkreis** [comment]
@bernardsilenou @ChristopherRiedel it might be differ from country to country how the user roles are used for. Why not make a configuration for the server if they are shown or not?

**Comment by MartinWahnschaffe** [comment]
@ChristopherRiedel Since the REST_USER is now using NATIONAL jurisdiction level (to have access to all data) it can no longer be combined with a district user. Please create an issue to replace the current mechanism in the survnet converter initial import that depends on the user's district.

**Comment by zohaibhassan122** [comment]
I am hoping that we are adding a Mobile app login right as well as configurable variable.

**Comment by MartinWahnschaffe** [comment]
I looked into the REST_USER topic again. 

The root problem is that we are no longer able to give the REST_USER access to data across all jurisdictions based on the user role (and not based on jurisdiction). We thought that giving the REST_USER the NATION jurisdiction would mitigate this.

The REST_USER user role only comes with the SORMAS_REST user right. This means it currently always has to be combined with another user role to have the right to access certain types of data. Based on that it does not make a lot of sense to change the jurisdiction level of the REST_USER to NATION, because it would introduce conflicts for all existing users that combine REST_USER with another role.

So instead I suggest to (discussed this with @ChristopherRiedel):

- [x] Remove the REST_USER user role. 
- [x] give the SORMAS_REST user right to all default user roles, so sormas angular and other external systems that rely on this don't face problems.
- [x] Communicate in the release notes that external systems that have relied on the REST_USER user role to get access to all data have to rely on using a NATIONAL user role instead. Those system have to touch the configuration of those users anyway.

We will keep the SORMAS_REST user right. Maybe there will be other interfaces to SORMAS in the future, so there is no harm in keeping this right.

**Comment by StefanKock** [comment]
Build is broken:
```bash
DashboardServiceTest.java:[15,33] cannot find symbol: class UserRole
````

**Comment by kwa20** [comment]
After the re-deployment with this commit, the test instances are no longer reachable, Independent of them having a local or keycloak login.

| test server | login | error |
|:-|:-|:-
|test.sormas.netzlink.com|local|login fails|
|test-auto.sormas.netzlink.com|local|login fails|
|test-de.sormas.netzlink.com|local|login fails|
|test-de1.sormas.netzlink.com|keycloak|403 Forbidden|
|test-de2.sormas.netzlink.com|keycloak|403 Forbidden|
|test-de3.sormas.netzlink.com|keycloak|400 Bad Gateway|

**Comment by adinaflorea9** [comment]
Mobile app: logged in with a Hospital Informant - the synchronization with the server is not successful; an error message is displayed in the mobile app:
![image](https://user-images.githubusercontent.com/96909231/170979908-d71b6565-c511-47ab-aba7-d81d6b0fd73a.png)

The logs from the local machine at the time when the synchronization was triggered and the error message was displayed in the app:
[logs hospital informant.txt](https://github.com/hzi-braunschweig/SORMAS-Project/files/8797833/logs.hospital.informant.txt)


Note: from what I was able to test so far, this could only be reproduced with the Hospital Informant (tried with a Surveillance Officer as well, but the sync worked as expected).

Reported the issue here, as I could not determine whether this was connected or not with #4461 and should be further investigated by a developer.  

Commit hash: 372bd3e89e21d45d6c518bee112364dac4955a47
Device: Windows 10 (vaadin-app); Android Studio Emulator - Pixel C (android-app)
Android version/Browser: Chrome (vaadin-app); Android 11 (android-app)
Server URL: local machine
Sormas Version: 1.73.0-SNAPSHOT(372bd3e)
User Role: Hospital Informant (vaadin-app + android app)

**Comment by adinaflorea9** [comment]
Exception thrown when a Surveillance Officer tries to create a task in the mobile app - this is not reproducible on the earlier versions of the app (1.72.1) therefore I am logging this behavior as part of this ticket (if further investigation shows that this is not connected to #4461 I will open a separate issue).

Description:
The Surveillance Officer has the TASK_CREATE = YES; when the Surveillance Officer clicks on the 'Create task' button in the 'Tasks' directory, or on a case or contact page, it's visible in the logs that an exception is thrown and the user is either redirected to the PIN page in the mobile app, or the app 'crashes' and a notification is displayed on the mobile device stating that the app is not responsive.

### Steps to Reproduce

1. Log into the mobile app with a Surveillance Officer;
2. Navigate to the 'Tasks' directory and click + to create a new task.

### Actual Behavior

The user is redirected to the PIN page or a notification is displayed stating that the app is not responsive; in the logs, it is visible that an exception is thrown.

If the user re-enters their PIN and enters the app, none of the directories are visible:
![image](https://user-images.githubusercontent.com/96909231/172330060-73cc4e2c-338f-47b2-a71e-7157400cdf77.png)


Screenshot of the exception from the local machine logs:
![image](https://user-images.githubusercontent.com/96909231/172329493-bcbe2cd3-d32b-48be-a386-cd9fe794164a.png)

### System Details

* Device:  Android Studio Emulator - Pixel C (android-app)
* SORMAS version: 1.73.0-SNAPSHOT(21bd3b8)
* Android version/Browser: Android 11 (android-app)
* Server URL: local machine
* User Role: Surveillance Officer

**Comment by MateStrysewske** [comment]
Bug found by @adinaflorea9 16 days ago has been explained in more detail in #9481.
